fs-storage/src/file_storage.rs

```diff
@@ -176,9 +157,77 @@ where
                 ),
             ));
         }
+
+        Ok(data)
+    }
+}
+
+impl<K, V> BaseStorage<K, V> for FileStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr,
+    V: Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + Monoid<V>,
+{
+    /// Set a key-value pair in the storage
+    fn set(&mut self, key: K, value: V) {
+        self.data.entries.insert(key, value);
+        self.modified = std::time::SystemTime::now();
+    }
+
+    /// Remove a key-value pair from the storage given a key
+    fn remove(&mut self, id: &K) -> Result<()> {
+        self.data.entries.remove(id).ok_or_else(|| {
+            ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
+        })?;
+        self.modified = std::time::SystemTime::now();
+        Ok(())
+    }
+
+    /// Compare the timestamp of the storage file
+    /// with the timestamp of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
+    fn needs_syncing(&self) -> Result<SyncStatus> {
+        let file_updated = fs::metadata(&self.path)?.modified()?;
+
+        // mapping updated
+        if self.modified > self.written_to_disk {
+            // file updated since last write
+            if file_updated > self.written_to_disk {
+                // both updated
+                return Ok(SyncStatus::FullSync);
+            } else {
+                // only mapping updated
+                return Ok(SyncStatus::DownSync);
+            }
+        } else {
+            // file updated since last write
+            if file_updated > self.written_to_disk {
+                // only file updated
+                return Ok(SyncStatus::UpSync);
+            } else {
+                // no change
+                return Ok(SyncStatus::NoSync);
+            }
```

Comment:
I'm sorry, but could you explain the logic here a bit more? I'm not sure I get it.

Reply:
Actually, previous `match` expression looked more readable to me.

@tareknaser Ishan introduced one more timestamp, so now we track times of both internal modification and writing to the filesystem. We have third timestamp in the filesystem itself. These 3 timestamps are enough to figure out which side updated, or if both sides been updated.